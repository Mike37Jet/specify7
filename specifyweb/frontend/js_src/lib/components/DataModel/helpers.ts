import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import { defined } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import type {
  AnySchema,
  AnyTree,
  SerializedModel,
  SerializedResource,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import { parseResourceUrl, resourceToJson } from './resource';
import { strictGetModel } from './schema';
import type { Tables } from './types';
import { addMissingFields } from './addMissingFields';

/** Like resource.toJSON(), but keys are converted to camel case */
export const serializeResource = <SCHEMA extends AnySchema>(
  resource: SerializedModel<SCHEMA> | SpecifyResource<SCHEMA>
): SerializedResource<SCHEMA> =>
  serializeModel<SCHEMA>(
    typeof resource.toJSON === 'function'
      ? resourceToJson(resource as SpecifyResource<SCHEMA>)
      : (resource as SerializedModel<SCHEMA>),
    (resource as SpecifyResource<SCHEMA>)?.specifyModel?.name
  );

export const specialFields = new Set([
  'id',
  'resource_uri',
  'recordset_info',
  '_tableName',
]);

/** Recursive helper for serializeResource */
function serializeModel<SCHEMA extends AnySchema>(
  resource: SerializedModel<SCHEMA>,
  tableName?: keyof Tables
): SerializedResource<SCHEMA> {
  const model = strictGetModel(
    defined(
      (tableName as SCHEMA['tableName']) ??
        resource._tableName ??
        parseResourceUrl((resource.resource_uri as string) ?? '')?.[0],
      `Unable to serialize resource because table name is unknown.` +
        (process.env.NODE_ENV === 'test'
          ? `\nMake sure your test file calls requireContext();`
          : '')
    )
  );
  const fields = [...model.fields.map(({ name }) => name), model.idField.name];

  return addMissingFields(
    model.name,
    Object.fromEntries(
      Object.entries(resource).map(([lowercaseFieldName, value]) => {
        let camelFieldName = fields.find(
          (fieldName) => fieldName.toLowerCase() === lowercaseFieldName
        );
        if (camelFieldName === undefined) {
          camelFieldName = lowercaseFieldName;
          if (!specialFields.has(lowercaseFieldName))
            console.warn(
              `Trying to serialize unknown field ${lowercaseFieldName} for table ${model.name}`,
              resource
            );
        }
        if (
          typeof value === 'object' &&
          value !== null &&
          !specialFields.has(camelFieldName)
        ) {
          const field = model.getField(lowercaseFieldName);
          const tableName =
            field === undefined || !field.isRelationship
              ? undefined
              : field.relatedModel.name;
          return [
            camelFieldName,
            Array.isArray(value)
              ? value.map((value) =>
                  serializeModel(
                    value as unknown as SerializedModel<SCHEMA>,
                    tableName
                  )
                )
              : serializeModel(value as SerializedModel<AnySchema>, tableName),
          ];
        } else return [camelFieldName, value];
      })
    )
  ) as SerializedResource<SCHEMA>;
}

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyModel.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyModel.name === tableName ? resource : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyModel.name) ? resource : undefined;
