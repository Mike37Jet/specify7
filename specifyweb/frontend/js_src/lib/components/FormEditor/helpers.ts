import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AttachmentOwnerBaseBusRules',
  'AttachmentReferenceBaseBusRules',
  'AuthorBusRules',
  'BaseTreeBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventAttributeBusRules',
  'CollectingEventAuthorizationBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectionSetupBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeaccessionBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DisposalAgentBusRules',
  'DisposalBusRules',
  'DisposalPreparationBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeInBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeoCoordDetailBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'InstitutionBusRules',
  'JournalBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'LocalityDetailBusRules',
  'PaleoContextBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'ReferenceWorkBusRules',
  'RepositoryAgreementBusRules',
  'ShipmentBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TableSearchResults',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeableSearchQueryBuilder',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
