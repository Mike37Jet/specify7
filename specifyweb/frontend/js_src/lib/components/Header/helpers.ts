export function formatDate(date: Date) {
  const year = date.getFullYear();
  const month = date.getMonth().toString();
  const day = date.getDate().toString();
  const hours = date.getHours().toString();
  const minutes = date.getMinutes().toString();
  const seconds = date.getSeconds().toString();

  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}
