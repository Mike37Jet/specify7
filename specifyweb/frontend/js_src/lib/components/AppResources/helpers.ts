import type { AppResourcesConformation } from './Aside';
import type { AppResources, AppResourcesTree } from './hooks';
import type { SpAppResource, SpViewSetObj } from '../DataModel/types';
import { f } from '../../utils/functools';
import type { SpecifyModel } from '../DataModel/specifyModel';
import type { RA } from '../../utils/types';
import { SerializedResource } from '../DataModel/helperTypes';

export const getAppResource = (
  resources: AppResources,
  model: SpecifyModel<SpAppResource | SpViewSetObj>,
  resourceId: number | undefined
): SerializedResource<SpAppResource | SpViewSetObj> | false | undefined =>
  typeof resourceId === 'number'
    ? (
        (model.name === 'SpAppResource'
          ? resources.appResources
          : resources.viewSets) as RA<SerializedResource<SpAppResource>>
      ).find(({ id }) => id === resourceId) ?? false
    : undefined;

export type AppResourceMode = 'appResources' | 'viewSets';

export const getAppResourceMode = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): AppResourceMode =>
  resource._tableName === 'SpAppResource' ? 'appResources' : 'viewSets';

export const getAppResourceCount = (
  resourcesTree: AppResourcesTree[number]
): number =>
  resourcesTree.appResources.length +
  resourcesTree.viewSets.length +
  f.sum(resourcesTree.subCategories.map(getAppResourceCount));

export const buildAppResourceConformation = (
  resourcesTree: AppResourcesTree
): RA<AppResourcesConformation> =>
  resourcesTree
    .filter(
      ({ subCategories, appResources, viewSets }) =>
        subCategories.length > 0 ||
        appResources.length > 0 ||
        viewSets.length > 0
    )
    .map(({ key, subCategories }) => ({
      key,
      children: buildAppResourceConformation(subCategories),
    }));
