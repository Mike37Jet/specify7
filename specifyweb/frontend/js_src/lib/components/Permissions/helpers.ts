import { getCache } from '../../utils/cache';
import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import { defined, filterArray } from '../../utils/types';
import type { AnyTree } from '../DataModel/helperTypes';
import { schema } from '../DataModel/schema';
import type { Tables } from '../DataModel/types';
import { toolDefinitions } from '../Security/registry';
import { tableNameToResourceName } from '../Security/utils';
import type { tableActions } from './definitions';
import {
  getDerivedPermissions,
  getOperationPermissions,
  getTablePermissions,
} from './index';

const isReadOnly = getCache('forms', 'readOnlyMode') === true;

// REFACTOR: use <ProtectedTable> and etc in favor of this function
/**
 * Security errors are logged so that admins can see why a particular UI
 * component is disabled or missing
 */
export function hasTablePermission(
  tableName: keyof Tables,
  action: typeof tableActions[number],
  collectionId = schema.domainLevelIds.collection
): boolean {
  if (isReadOnly && action !== 'read') return false;
  if (
    defined(getTablePermissions())[collectionId][
      tableNameToResourceName(tableName)
    ][action]
  )
    return true;
  console.log(`No permission to ${action} ${tableName}`);
  return false;
}

export const hasPermission = <
  RESOURCE extends keyof ReturnType<typeof getOperationPermissions>[number]
>(
  resource: RESOURCE,
  action: keyof ReturnType<typeof getOperationPermissions>[number][RESOURCE],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  defined(getOperationPermissions())[collectionId][resource][action]
    ? true
    : f.log(`No permission to ${action.toString()} ${resource}`) ?? false;

export const hasToolPermission = (
  tool: keyof ReturnType<typeof toolDefinitions>,
  action: typeof tableActions[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  (toolDefinitions()[tool].tables as RA<keyof Tables>).every((tableName) =>
    hasTablePermission(tableName, action, collectionId)
  );

export const hasTreeAccess = (
  treeName: AnyTree['tableName'],
  action: typeof tableActions[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  hasTablePermission(treeName, action, collectionId) &&
  hasTablePermission(`${treeName}TreeDef`, action, collectionId) &&
  hasTablePermission(`${treeName}TreeDefItem`, action, collectionId);

export const hasDerivedPermission = <
  RESOURCE extends keyof ReturnType<typeof getDerivedPermissions>[number]
>(
  resource: RESOURCE,
  action: keyof ReturnType<typeof getDerivedPermissions>[number][RESOURCE],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  defined(getDerivedPermissions())[collectionId][resource][action]
    ? true
    : f.log(`No permission to ${action.toString()} ${resource}`) ?? false;

/** Check if user has a given permission for each table in a mapping path */
export const hasPathPermission = (
  baseTableName: keyof Tables,
  mappingPath: RA<string>,
  action: typeof tableActions[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  mappingPathToTableNames(baseTableName, mappingPath, true).every((tableName) =>
    hasTablePermission(tableName, action, collectionId)
  );

export const mappingPathToTableNames = (
  baseTableName: keyof Tables,
  mappingPath: RA<string>,
  ignoreBaseTable = false
): RA<keyof Tables> =>
  f.unique(
    filterArray(
      mappingPath.flatMap((_, index) => {
        if (index === 0 && ignoreBaseTable) return undefined;
        const field = schema.models[baseTableName].getField(
          mappingPath.slice(index).join('.')
        );
        return typeof field === 'object'
          ? [
              field.model.name,
              field.isRelationship ? field.relatedModel.name : undefined,
            ]
          : undefined;
      })
    )
  );
