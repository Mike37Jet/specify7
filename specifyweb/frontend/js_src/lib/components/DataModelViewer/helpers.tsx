import React from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { schemaText } from '../../localization/schema';
import { syncFieldFormat } from '../../utils/fieldFormat';
import { f } from '../../utils/functools';
import { resolveParser } from '../../utils/parser/definitions';
import { ensure, IR, RA, RR } from '../../utils/types';
import { formatNumber } from '../Atoms/Internationalization';
import { getField } from '../DataModel/helpers';
import { schema } from '../DataModel/schema';
import { TableIcon } from '../Molecules/TableIcon';

export type Value =
  | number
  | string
  | readonly [number | string | undefined, JSX.Element]
  | undefined;
export type Row<SHAPE extends IR<Value>> = SHAPE;

const parser = f.store(() =>
  resolveParser(
    {},
    {
      type: 'java.lang.Boolean',
    }
  )
);

export const booleanFormatter = (value: boolean): string =>
  syncFieldFormat(undefined, parser(), value);

export const tableColumns = f.store(
  () =>
    ({
      name: getField(schema.models.SpLocaleContainer, 'name').label,
      label: schemaText.fieldLabel(),
      isSystem: getField(schema.models.SpLocaleContainer, 'isSystem').label,
      isHidden: getField(schema.models.SpLocaleContainer, 'isHidden').label,
      tableId: schemaText.tableId(),
      fieldCount: schemaText.fieldCount(),
      relationshipCount: schemaText.relationshipCount(),
    } as const)
);

export const getTables = () =>
  ensure<RA<Row<RR<keyof ReturnType<typeof tableColumns>, Value>>>>()(
    Object.values(schema.models).map(
      (model) =>
        ({
          name: [
            model.name.toLowerCase(),
            <>
              <TableIcon label={false} name={model.name} />
              {model.name}
            </>,
          ],
          label: model.label,
          isSystem: booleanFormatter(model.isSystem),
          isHidden: booleanFormatter(model.isHidden),
          tableId: [
            model.tableId,
            <span className="flex w-full justify-end tabular-nums">
              {model.tableId}
            </span>,
          ],
          fieldCount: [
            model.fields.length,
            <span className="flex w-full justify-end tabular-nums">
              {formatNumber(model.fields.length)}
            </span>,
          ],
          relationshipCount: [
            model.relationships.length,
            <span className="flex w-full justify-end tabular-nums">
              {formatNumber(model.relationships.length)}
            </span>,
          ],
        } as const)
    )
  );

export function DataModelRedirect(): null {
  const { tableName = '' } = useParams();
  const navigate = useNavigate();
  React.useEffect(() => {
    navigate(`/specify/data-model/#${tableName}`, { replace: true });
  }, []);
  return null;
}
