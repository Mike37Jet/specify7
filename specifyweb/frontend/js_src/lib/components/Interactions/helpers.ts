import { ajax } from '../../utils/ajax';
import { formData } from '../../utils/ajax/helpers';
import type { RA } from '../../utils/types';

export type Preparations = RA<{
  readonly catalogNumber: string;
  readonly taxon: string;
  readonly preparationId: number;
  readonly prepType: string;
  readonly countAmount: number;
  readonly loaned: number;
  readonly gifted: number;
  readonly exchanged: number;
  readonly available: number;
}>;

export type PreparationRow = readonly [
  string,
  string,
  number,
  string,
  number,
  string | null,
  string | null,
  string | null,
  string
];
export const getPrepsAvailableForLoanRs = async (recordSetId: number) =>
  ajax<RA<PreparationRow>>(
    `/interactions/preparations_available_rs/${recordSetId}/`,
    {
      headers: { Accept: 'application/json' },
    }
  ).then(({ data }) => data);

export const getPrepsAvailableForLoanCoIds = async (
  idField: string,
  collectionObjectIds: RA<string>
) =>
  ajax<RA<PreparationRow>>('/interactions/preparations_available_ids/', {
    method: 'POST',
    headers: { Accept: 'application/json' },
    body: formData({
      id_fld: idField,
      co_ids: collectionObjectIds,
    }),
  }).then(({ data }) => data);

export const getInteractionsForPrepId = async (prepId: number) =>
  ajax<RA<readonly [number, string | null, string | null, string | null]>>(
    '/interactions/prep_interactions/',
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ prepIds: prepId }),
    }
  ).then(({ data }) => data[0]);

export const getPrepAvailability = async (
  prepId: number,
  iPrepId: number | undefined,
  iPrepName: string
) =>
  ajax(
    `/interactions/prep_availability/${prepId}/${
      typeof iPrepId === 'number' ? `${iPrepId}/${iPrepName}/` : ''
    }`,
    { headers: { Accept: 'application/json' } }
  ).then(({ data }) => data);
