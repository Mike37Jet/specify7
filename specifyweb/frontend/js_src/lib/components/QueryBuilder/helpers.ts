import { getTransitionDuration } from '../UserPreferences/Hooks';
import type { QueryFieldFilter } from './FieldFilter';
import { queryFieldFilters } from './FieldFilter';
import type { MappingPath } from '../WbPlanView/Mapper';
import type { SpQueryField, Tables } from '../DataModel/types';
import type { SerializedResource } from '../DataModel/helpers';
import { group, KEY, removeKey, sortFunction, VALUE } from '../../utils/utils';
import { QueryFieldSpec } from './fieldSpec';
import type { RA } from '../../utils/types';
import { defined } from '../../utils/types';
import { parserFromType, parseValue } from '../../utils/uiParse';
import { mappingPathToString } from '../WbPlanView/mappingHelpers';
import type { MappingLineData } from '../WbPlanView/navigator';
import { mappingPathIsComplete } from '../WbPlanView/helpers';

export type SortTypes = 'ascending' | 'descending' | undefined;
export const sortTypes: RA<SortTypes> = [undefined, 'ascending', 'descending'];
export const flippedSortTypes = {
  none: 0,
  ascending: 1,
  descending: 2,
} as const;

export type QueryField = {
  /*
   * ID is used only as a React [key] prop in order to optimize rendering when
   * query lines move
   */
  readonly id: number;
  readonly mappingPath: MappingPath;
  readonly sortType: SortTypes;
  readonly isDisplay: boolean;
  readonly filters: RA<{
    readonly type: QueryFieldFilter;
    readonly startValue: string;
    readonly isNot: boolean;
  }>;
};

/** Convert SpQueryField to internal QueryField representation */
export function parseQueryFields(
  queryFields: RA<SerializedResource<SpQueryField>>
): RA<QueryField> {
  return group(
    Array.from(queryFields)
      .sort(sortFunction(({ position }) => position))
      .map(({ isNot, isDisplay, ...field }, index) => {
        const fieldSpec = QueryFieldSpec.fromStringId(
          field.stringId,
          field.isRelFld ?? false
        );

        /*
         * SpQueryField.startValue containing fullDate may be in different formats
         * See https://github.com/specify/specify7/issues/1348
         */
        const startValue =
          typeof field.startValue === 'string' &&
          fieldSpec.datePart === 'fullDate'
            ? field.startValue
                .split(',')
                .map((value) =>
                  parseValue(
                    parserFromType('java.sql.Timestamp'),
                    undefined,
                    value
                  )
                )
                .map((parsed) =>
                  parsed?.isValid
                    ? (parsed.parsed as string)
                    : field.startValue ?? ''
                )
                .join(',')
            : field.startValue;

        const mappingPath = fieldSpec.toMappingPath();
        return [
          mappingPathToString(mappingPath),
          {
            id: index,
            mappingPath,
            sortType: sortTypes[field.sortType],
            filter: {
              type: defined(
                Object.entries(queryFieldFilters).find(
                  ([_, { id }]) => id === field.operStart
                )
              )[KEY],
              isNot,
              startValue,
            },
            isDisplay,
          },
        ] as const;
      })
  ).map(([_mappingPath, groupedFields]) => ({
    ...removeKey(groupedFields[0], 'filter'),
    filters: groupedFields.map(({ filter }) => filter),
  }));
}

export const queryFieldsToFieldSpecs = (
  baseTableName: keyof Tables,
  fields: RA<QueryField>
): RA<readonly [QueryField, QueryFieldSpec]> =>
  fields
    .filter(({ mappingPath }) => mappingPathIsComplete(mappingPath))
    .map((field) => [
      field,
      QueryFieldSpec.fromPath(baseTableName, field.mappingPath),
    ]);

const auditLogMappingPaths = [
  ['parentRecordId'],
  ['parentTableNum'],
  ['recordId'],
  ['tableNum'],
  ['fields', 'oldValue'],
  ['fields', 'newValue'],
  ['fields', 'fieldName'],
];

/** Add audit log fields required to format audit log results (if missing) */
export const addAuditLogFields = (
  baseTableName: keyof Tables,
  fields: RA<QueryField>
): RA<QueryField> =>
  baseTableName === 'SpAuditLog'
    ? [
        ...fields.map((field) => ({
          ...field,
          // Force display audit log fields
          isDisplay:
            field.isDisplay ||
            auditLogMappingPaths.some(
              (mappingPath) =>
                mappingPathToString(field.mappingPath) ===
                mappingPathToString(mappingPath)
            ),
        })),
        // Add missing audit log fields
        ...auditLogMappingPaths
          .filter((mappingPath) =>
            fields.every(
              (field) =>
                mappingPathToString(field.mappingPath) !==
                mappingPathToString(mappingPath)
            )
          )
          .map(
            (mappingPath) =>
              ({
                id: 0,
                mappingPath,
                sortType: undefined,
                isDisplay: true,
                parser: undefined,
                filters: [
                  {
                    type: 'any',
                    startValue: '',
                    isNot: false,
                  },
                ],
              } as const)
          ),
      ]
    : fields;

/** Convert internal QueryField representation to SpQueryFields */
export const unParseQueryFields = (
  baseTableName: keyof Tables,
  fields: RA<QueryField>
): RA<SerializedResource<SpQueryField>> =>
  queryFieldsToFieldSpecs(baseTableName, fields).flatMap(
    ([field, fieldSpec], index) => {
      const commonData = {
        ...fieldSpec.toSpQueryAttributes(),
        sortType: sortTypes.indexOf(field.sortType),
        position: index,
        isDisplay: field.isDisplay,
      };

      const hasFilters = field.filters.some(({ type }) => type !== 'any');
      return field.filters
        .filter((filter, index) =>
          // Filter out duplicate or redundant "any"
          hasFilters ? filter.type !== 'any' : index === 0
        )
        .map(
          ({ type, startValue, isNot }, index) =>
            ({
              ...commonData,
              operStart: defined(
                // Back-end treats "equal" with blank startValue as "any"
                Object.entries(queryFieldFilters).find(
                  ([name]) => name === type
                )
              )[VALUE].id,
              startValue,
              isNot,
              /*
               * Prevent OR conditions from returning separate column in the
               * results
               */
              isDisplay: commonData.isDisplay && index === 0,
              // REFACTOR: add missing nullable fields here
            } as unknown as SerializedResource<SpQueryField>)
        );
    }
  );

export function hasLocalityColumns(fields: RA<QueryField>): boolean {
  const fieldNames = new Set(
    fields
      .filter(({ isDisplay }) => isDisplay)
      .map(({ mappingPath }) => mappingPath.at(-1))
  );
  return fieldNames.has('latitude1') && fieldNames.has('longitude1');
}

export const mutateLineData = (
  lineData: RA<MappingLineData>
): RA<MappingLineData> =>
  lineData.filter(
    ({ customSelectSubtype }) => customSelectSubtype !== 'toMany'
  );

export function smoothScroll(element: HTMLElement, top: number): void {
  if (typeof element.scrollTo === 'function')
    element.scrollTo({
      top,
      behavior: getTransitionDuration() === 0 ? 'auto' : 'smooth',
    });
  else element.scrollTop = element.scrollHeight;
}
