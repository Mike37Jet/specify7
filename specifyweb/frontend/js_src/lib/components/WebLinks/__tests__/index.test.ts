import { requireContext } from '../../../tests/helpers';
import { webLinks } from '../index';

requireContext();

test('weblinks are fetched and parsed correctly', async () =>
  expect(
    webLinks.then((webLinks) =>
      Object.fromEntries(
        webLinks.map((webLink) => [webLink.name, webLink] as const)
      )
    )
  ).resolves.toMatchSnapshot());
