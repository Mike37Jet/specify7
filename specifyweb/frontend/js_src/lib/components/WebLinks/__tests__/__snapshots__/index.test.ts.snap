// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weblinks are fetched and parsed correctly 1`] = `
{
  "AccessionNumber": {
    "args": [
      "accessionNumber",
    ],
    "description": "",
    "isExternal": false,
    "template": "http://specify6-test.nhm.ku.edu/echo.php?text1=<%= accessionNumber %>",
  },
  "CTScan": {
    "args": [
      "_this",
    ],
    "description": "",
    "isExternal": false,
    "template": "https://www.morphosource.org/biological_specimens/<%= _this %>",
  },
  "FishBase": {
    "args": [
      "genus",
      "species",
    ],
    "description": "Displays a FishBase page for the Genus and Species",
    "isExternal": false,
    "template": "http://www.fishbase.org/Summary/<%= genus %>-<%= species %>.html",
  },
  "GUID": {
    "args": [
      "_this",
    ],
    "description": "",
    "isExternal": false,
    "template": "urn:uuid:<%= _this %>",
  },
  "Genbank": {
    "args": [
      "_this",
    ],
    "description": "",
    "isExternal": false,
    "template": "http://www.ncbi.nlm.nih.gov/nuccore/<%= _this %>",
  },
  "MailTo": {
    "args": [
      "email",
    ],
    "description": "Launches an email message window for sending an email.",
    "isExternal": false,
    "template": "mailto:<%= email %>",
  },
  "Specify7": {
    "args": [
      "catalogNumber",
    ],
    "description": "",
    "isExternal": false,
    "template": "https://ichthyology.specify.ku.edu/specify/bycatalog/KUI/<%= catalogNumber %>/",
  },
  "WEBLINK": {
    "args": [
      "_this",
    ],
    "description": "",
    "isExternal": false,
    "template": "http://specify6-test.nhm.ku.edu/echo.php?text1=<%= _this %>",
  },
}
`;
